ls1intum/Artemis
# [Exam] Client enhancements for programming exercises

### Checklist

### Motivation and Context
After discussing the code editor functionality in length with Stephan and the assignees, we decided to remove the ability to work offline with the code-editor, as the implementation grew in complexity. This continues to be an interesting feature but requires an in-depth analysis and code refactoring. For now, the students who want to work offline will have to clone the repository. To this sense, we want to update the disconnected notification for the user, so that he is aware of its effects on his changes. Furthermore, two bugs we found, one where the exercise instructions were not loaded while the user is offline, and the exercises not being saved, when the user switches between exercises. 

### Description
- [x] Add an "Online Editor" heading to emphasise the **online** part
- [x] In case the user is disconnected display different notifications depending on: programming exercise && codeEditor && OfflineIde/ noOfflineIde
- [x] Display the instructions if offline
- [x] Save and commit changes when switching to another exercise

### Steps for Testing

1. Create a new exam with three exercise groups. Create three programming exercises with different onlineCodeEditor and offlineIDE options. Add students, generate the student exams and participations.
2. Start the exam
3. Disconnect from the internet (if you test locally, stop the server or check if your browser has an offline mode  e.g. chrome -> dev tools -> network tab -> change to offline)
4. Switch between the exercises
--> Instructions should still be shown
--> The footer should display an appropriate message depending on the options codeEditorEnabled & OfflineIde
5. Go online
6. Switch to a programming exercise with online code editor enabled
7. Make changes
8. Switch to another exercise
9. Switch back
--> Changes should be saved
--> Build should run or the results should be already up-to-date

### Screenshots
Saving and committing changes on exercise change: 
![Commit afer exercise change](https://user-images.githubusercontent.com/15110960/86060325-cb634d00-ba64-11ea-8111-ade243d23a56.gif)

#### Footer changes (also notice Code editor title + instructions being loaded offline)
- Code Editor Only:
![Code Editor Only](https://user-images.githubusercontent.com/15110960/86060209-935c0a00-ba64-11ea-8c2c-de5444564f2b.PNG)

- Code Editor + clone button
![Code editor plus offline](https://user-images.githubusercontent.com/15110960/86060265-acfd5180-ba64-11ea-8266-3bb50d4f6f2f.PNG)

- Offline Ide 
![offline](https://user-images.githubusercontent.com/15110960/86060290-b7b7e680-ba64-11ea-8037-0c4beaf0bdba.PNG)
![online offline](https://user-images.githubusercontent.com/15110960/86060300-bc7c9a80-ba64-11ea-9fbf-bb2579537e46.PNG)
